text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,248 @@
+cannot be assumed, and in many cases it is not even desired, that all those
+components and layers are instrumented and propagate context according to
+OpenTelemetry requirements.
+
+A _creation context_ allows correlating the producer with the consumers of a
+message, regardless of intermediary instrumentation. The creation context is
+created by the producer and must be propagated to the consumers. It must not be
+altered by intermediaries.  This context helps to model dependencies between
+producers and consumers, regardless of the underlying messaging transport
+mechanism and its instrumentation.
+
+Instrumentors are required to instrument producer and consumer applications
+so that context is attached to messages and extraced from messages in a
+coordinated way. Future versions of these conventions might recommend [context propagation according to certain industry standards](#standards-for-context-propagation).
+
+> A producer SHOULD attach a creation context to each message. The creation context
+> SHOULD be attached in a way so that it is not changed by intermediaries.
+
+### Trace structure, names, and attributes
+
+#### Consumer
+
+For many use cases, it is not possible to rely on the presence of "Process"
+spans for correlating producer with consumer traces: there are cases where a
+dedicated processing operation cannot be identified, or where processing
+happens in a different trace. Furthermore, processing operations often are not
+covered by messaging libraries and SDKs, but take place in application code.
+Consistently creating spans for "Processing" operations would require either
+effort from the application owner to correctly instrument those operations, or
+additional capabilities of messaging libraries and SDKs (e. g. hooks for
+processing callbacks, which can then be instrumented by the libraries or SDKs).
+
+While it is possible to create "Process" spans and correlate those with
+consumer traces in certain cases, this is not something that can be generally
+required. Therefore, it is more feasible to require the creation of "Deliver"
+spans (for push-based APIs) or "Receive" spans (for pull-based APIs) to
+correlate producer with consumer traces.
+
+##### Instrumenting push-based scenarios
+
+In push-based consumer scenarios, the delivery of messages is not initiated by
+the application code. Instead, callbacks or handlers are registered and then
+called by messaging SDKs to forward messages to the application.
+
+A "Deliver" span covers the call of such a callback or handler, and should link
+to the "Create" spans of all messages that are forwarded via the respective
+call. Depending on the use case, "Deliver" spans can correlate with "Process"
+spans or other spans modelling processing operations.
+
+> "Deliver" spans SHOULD be created for operations of passing messages to the
+> application, when those operations are not initiated by the application
+> code.
+
+##### Instrumenting pull-based scenarios
```

Comment:
my take on it is:
There are 2 different dimensions here: push-pull and callback-return patterns.

The push-pull is mainly about the ambient context and history. 

### push
the consume trace (in the consumer service) has no "history" (within the process) or ambient context. it is the first operation in that service and thus a bit simpler than pull scenario. It is the direct continuation of the operations that happened prior to it in the intermediator and can potentially be connected with parent-child relationship and derive sampling decisions.

### pull
The consumer trace usually starts with some sort of an RPC request from the service's business logic and can potentially have a history at the time messages are consumed. It can also be somehow reasonable to expect 2 different spans in this case - one describing the user's request (RPC operation invoked by user) and one describing the handling of those messages within the application's code.

---

The callback-return is about how messages are delivered back to the application and how the handling is scoped (active span + duration)

### callback
This pattern makes auto instrumentation easy as the handling of messages is well scoped in term of span activation and duration which are (somehow) well defined. callback pattern can be used in both push and pull mechanisims (for example, SQS in nodejs is pull based and has a callback mode).

### return
This is a complex scenario where the delivered messages handling is not semantically scoped by the program. To my knowledge and understanding, this can only happen with "pull" based systems.

Reply:
Good point.

I think for the vast majority of cases, the combinations are push-callback and pull-return.

The SQS case looks like pull-callback, but I think it can also be seen as a corner case of push-callback, in the sense that the application has no idea of when the supplied callback is called. In the SQS case, you might create a "Receive" span for the call to `receiveMessage`, and a "Deliver" span for the duration of the callback. However, the "Receive" span has most likely ended when the "Deliver" span starts. So, in this special case it makes maybe sense to start a new trace for the "Deliver" span (according to the push-callback model), and link it to the `receiveMessage` operation that triggered it?

Maybe we need some experimentation here. But as far as I understand, the SQS case can be modelled as a special case of the push-callback scenario.